Add the documentation for ReliableMessageListener [API-2182] (#941)
hazelcast/hz-docs: docs/modules/events/pages/object-events.adoc

        private long lastSequence;

        public void onMessage(Message<Long> m) {
            System.out.println("Received: " + m.getMessageObject());
        }

        @Override
        public long retrieveInitialSequence() {
            // The initial sequence to start reading from.
            return 1;
        }

        @Override
        public void storeSequence(long l) {
            // The sequence to store somewhere so that it can be retrieved upon restart.
            lastSequence = l;
            System.out.println("Stored sequence: " + l);
        }

        @Override
        public boolean isLossTolerant() {
            System.out.println("isLossTolerant called");
            // If true, the listener will not be removed upon an exception.
            return false;
        }

        @Override
        public boolean isTerminal(Throwable throwable) {
            System.out.println("isTerminal called");
            return false;
        }

        @Override
        public void onCancel() {
            System.out.println("onCancel called. The listener is being removed.");
        }
    }
----

**Durable Subscription**

> The ReliableMessageListener allows you to control where you want to start processing a message when the listener is registered. This makes it possible to create a durable subscription by storing the sequence of the last message and using this sequenceId as the sequenceId to start from.

Author: Changed to:

You can use the `ReliableMessageListener` to control which message you want processing to start when the listener is registered. This supports the creation of a durable subscription by storing the sequence of the last message and then re-starting from this `sequenceId`.

**Exception Handling**

> The ReliableMessageListener also gives the ability to deal with exceptions using the `isTerminal(Throwable)` method. This method allows you to control which exceptions should terminate the execution of the listener and cancel it. If a plain MessageListener is used, it won't terminate on exceptions and it will keep on running. But in some cases it is better to stop running.

Author: Changed as:

The `ReliableMessageListener` also copes with exceptions using the `isTerminal(Throwable)` method. This method allows you to control which exceptions can terminate and cancel the listener. If a `MessageListener` is used, it doesn't terminate following an exception and continues to run. In some situations, such as cluster being stopped, it is better to cancel the listener following the exception.

**Global Order**

The ReliableMessageListener will always get all events in order (global order). It will not get duplicates and there will only be gaps (loss of messages) if it is too slow. For more information see `isLossTolerant()`.

**Delivery Guarantees**

> Because the ReliableMessageListener controls which item it wants to continue from upon restart, it is very easy to provide an at-least-once or at-most-once delivery guarantee. The `storeSequence(long sequence)` is always called before a message is processed; so it can be persisted on some non-volatile storage. When the `retrieveInitialSequence()` returns the stored sequence, then an at-least-once delivery is implemented since the same item is now being processed twice. To implement an at-most-once delivery guarantee, add 1 to the stored sequence when the `retrieveInitialSequence()` is called.

Reviewer: ```suggestion
The `ReliableMessageListener` controls the item to continue from following a restart, providing an at-least-once or at-most-once delivery guarantee. The `storeSequence(long sequence)` is always called before a message is processed; so it can be persisted on non-volatile storage. When the `retrieveInitialSequence()` returns the stored sequence, an at-least-once delivery is implemented as the same item is processed again. To implement an at-most-once delivery guarantee, add 1 to the stored sequence when the `retrieveInitialSequence()` is called.
```

**Loss Tolerance**

> You can provide the `boolean isLosstolerant()` method return true if this ReliableMessageListener is able to deal with message loss. Even though the reliable topic promises to be reliable, it can be that a MessageListener is too slow. Eventually the message won't be available anymore.

Reviewer: ```suggestion
If this `ReliableMessageListener` can deal with message loss, the `boolean isLosstolerant()` method retursn` true`. Even though the reliable topic promises to be reliable, a `MessageListener` can be too slow, and eventually the message would become unavailable.
```

> If the ReliableMessageListener is not loss tolerant and the topic detects that there are missing messages, it will terminate the ReliableMessageListener.

Reviewer: ```suggestion
If the `ReliableMessageListener` is not loss tolerant and the topic detects that there are missing messages, the `ReliableMessageListener` terminates.
```

**onCancel Callback**

> This method is called by Hazelcast when the ReliableMessageListener is cancelled. This can happen when the listener is unregistered or cancelled due to an exception or during shutdown.

Reviewer: ```suggestion
This method is called by Hazelcast when the `ReliableMessageListener` is cancelled. This can happen when the listener is unregistered, or when it has been cancelled due to an exception or shutdown.
```

}
----

[[reliable-message-listener]]
== ReliableMessageListener

link:https://github.com/hazelcast/hazelcast/blob/master/hazelcast/src/main/java/com/hazelcast/topic/ReliableMessageListener.java[ReliableMessageListener] is a link:https://github.com/hazelcast/hazelcast/blob/master/hazelcast/src/main/java/com/hazelcast/topic/MessageListener.java[MessageListener] to better integrate with the reliable topic.

> If a regular MessageListener is registered on a reliable topic, it can't do much more than listening to new messages but the `ReliableMessageListener` gives you more capabilities on controlling the listener behaviour.

Reviewer: ```suggestion
If a standard `MessageListener` is registered on a reliable topic, it acts only as an asynchronous event handler for messages; however, a `ReliableMessageListener` allows you to control the behavior of the listener.
```

> If a `ReliableMessageListener` is registered on a normal ITopic, only the methods inherited from `MessageListener` are called.

Reviewer: ```suggestion
If a `ReliableMessageListener` is registered on a normal `ITopic`, only the methods inherited from `MessageListener` are called.
```

> The following is an example Reliable Message Listener class.

Reviewer: ```suggestion
For example, a Reliable Message Listener class can be defined as follows:
```

        private long lastSequence;

        public void onMessage(Message<Long> m) {
            System.out.println("Received: " + m.getMessageObject());
        }

        @Override
        public long retrieveInitialSequence() {
            // The initial sequence to start reading from.
            return 1;
        }

        @Override
        public void storeSequence(long l) {
            // The sequence to store somewhere so that it can be retrieved upon restart.
            lastSequence = l;
            System.out.println("Stored sequence: " + l);
        }

        @Override
        public boolean isLossTolerant() {
            System.out.println("isLossTolerant called");
            // If true, the listener will not be removed upon an exception.
            return false;
        }

        @Override
        public boolean isTerminal(Throwable throwable) {
            System.out.println("isTerminal called");
            return false;
        }

        @Override
        public void onCancel() {
            System.out.println("onCancel called. The listener is being removed.");
        }
    }
----

**Durable Subscription**

> You can use the `ReliableMessageListener` to control which message you want processing to start when the listener is registered. This supports the creation of a durable subscription by storing the sequence of the last message and then re-starting from this `sequenceId`.

Reviewer: ```suggestion
You can use the `ReliableMessageListener` to control which message to start processing from when the listener is registered. This supports the creation of a durable subscription by storing the sequence of the last message and then re-starting from the specified `sequenceId`.
```

**Exception Handling**

The `ReliableMessageListener` also copes with exceptions using the `isTerminal(Throwable)` method. This method allows you to control which exceptions can terminate and cancel the listener. If a `MessageListener` is used, it doesn't terminate following an exception and continues to run. In some situations, such as cluster being stopped, it is better to cancel the listener following the exception.

**Global Order**

> The `ReliableMessageListener` always gets all events in order (global order). It does not get duplicates and gaps (loss of messages) only occur when it is too slow. For more information on dealing with message loss, refer to the `isLossTolerant()` method in the https://docs.hazelcast.org/docs/5.4.0-SNAPSHOT/javadoc/com/hazelcast/topic/ReliableMessageListener.html#isLossTolerant()[Java API documentation^].

Reviewer: ```suggestion
The `ReliableMessageListener` always gets all events in order (global order). It does not get duplicates, and gaps (loss of messages) only occur when it is too slow. For more information on dealing with message loss, refer to the `isLossTolerant()` method in the https://docs.hazelcast.org/docs/5.4.0-SNAPSHOT/javadoc/com/hazelcast/topic/ReliableMessageListener.html#isLossTolerant()[Java API documentation^].
```

**Delivery Guarantees**

The `ReliableMessageListener` controls the item to continue from following a restart, providing an at-least-once or at-most-once delivery guarantee. The `storeSequence(long sequence)` is always called before a message is processed; so it can be persisted on non-volatile storage. When the `retrieveInitialSequence()` returns the stored sequence, an at-least-once delivery is implemented as the same item is processed again. To implement an at-most-once delivery guarantee, add 1 to the stored sequence when the `retrieveInitialSequence()` is called.

**Loss Tolerance**

> If this `ReliableMessageListener` can deal with message loss, the `boolean isLosstolerant()` method retursn` true`. Even though the reliable topic promises to be reliable, a `MessageListener` can be too slow, and eventually the message would become unavailable.

Reviewer: ```suggestion
If this `ReliableMessageListener` can deal with message loss, the `boolean isLosstolerant()` method returns `true`. Even though the reliable topic promises to be reliable, a `MessageListener` can be too slow and eventually this causes the message to become unavailable.
```

If the `ReliableMessageListener` is not loss tolerant and the topic detects that there are missing messages, the `ReliableMessageListener` terminates.

**onCancel Callback**

> This method is called by Hazelcast when the `ReliableMessageListener` is cancelled. This can happen when the listener is unregistered, or when it has been cancelled due to an exception or shutdown.

Reviewer: ```suggestion
This method is called by Hazelcast when the `ReliableMessageListener` is canceled. This can happen when the listener is unregistered, or when it has been canceled due to an exception or shutdown.
```
